pkg/util/rbacs.go

```diff
@@ -5,61 +5,76 @@ import (
 // will use this service account will fail to start since there are no credentials to pull any images.
-func CreateRBACs(ctx context.Context, sa *coreapi.ServiceAccount, role *rbacapi.Role, roleBindings []rbacapi.RoleBinding, client ctrlruntimeclient.Client, retryDuration, timeout time.Duration) error {
+func CreateRBACs(ctx context.Context, sa *coreapi.ServiceAccount, role *rbacapi.Role,
+	roleBindings []rbacapi.RoleBinding, client ctrlruntimeclient.Client,
+	retryDuration, timeout time.Duration) (*coreapi.ServiceAccount, error) {
 	skipPoll := false
 
 	err := client.Create(ctx, sa)
 	if err != nil && !kerrors.IsAlreadyExists(err) {
-		return results.ForReason("creating_service_account").WithError(err).Errorf("could not create service account %q: %v", sa.Name, err)
+		return nil, results.ForReason("creating_service_account").WithError(err).Errorf("could not create service account %q: %v", sa.Name, err)
 	}
 
 	if kerrors.IsAlreadyExists(err) {
 		skipPoll = true
 	}
 
 	if err := client.Create(ctx, role); err != nil && !kerrors.IsAlreadyExists(err) {
-		return results.ForReason("creating_roles").WithError(err).Errorf("could not create role %q: %v", role.Name, err)
+		return nil, results.ForReason("creating_roles").WithError(err).Errorf("could not create role %q: %v", role.Name, err)
 	}
 	for _, roleBinding := range roleBindings {
 		if err := client.Create(ctx, &roleBinding); err != nil && !kerrors.IsAlreadyExists(err) {
-			return results.ForReason("binding_roles").WithError(err).Errorf("could not create role binding %q: %v", roleBinding.Name, err)
+			return nil, results.ForReason("binding_roles").WithError(err).Errorf("could not create role binding %q: %v", roleBinding.Name, err)
 		}
 	}
 
 	if skipPoll {
-		return nil
+		return sa, nil
+	}
+
+	hasDockerCfgImagePullSecretSet := func(imagePullSecrets []coreapi.LocalObjectReference) bool {
+		for _, secret := range imagePullSecrets {
+			if api.RegistryPullCredentialsSecret != secret.Name {
+				return true
+			}
+		}
+		return false
 	}
 
+	var actualSa *coreapi.ServiceAccount
 	if err := wait.PollImmediate(retryDuration, timeout, func() (bool, error) {
-		actualSA := &coreapi.ServiceAccount{}
+		actualSa = &coreapi.ServiceAccount{}
 		if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{
 			Namespace: sa.Namespace,
 			Name:      sa.Name,
-		}, actualSA); err != nil {
+		}, actualSa); err != nil {
 			return false, fmt.Errorf("couldn't get service account %s: %w", sa.Name, err)
 		}
 
-		if len(actualSA.ImagePullSecrets) == 0 {
+		if !hasDockerCfgImagePullSecretSet(actualSa.ImagePullSecrets) {
 			return false, nil
 		}
 
 		return true, nil
-	},
-	); err != nil {
-		return results.ForReason("create_dockercfg_secrets").WithError(err).Errorf("timeout while waiting for dockercfg secret creation for service account %q: %v", sa.Name, err)
```

Comment:
I would not remove this line, just do not return here.

Reply:
What is it useful for if we are not going to return it?

Reply:
`results.()` sends data to result-aggregator and then prometheus.